### PR TITLE
Read allocated subnet CIDRs into controllercontext AZs

### PR DESCRIPTION
**Pull request by tuommaki**
Present AZ-Subnet CIDR pairs must be kept unchanged when allocating resources
for new AZs. Otherwise CloudFormation would tear down resources and break
connected NodePools as well.

### CODE REVIEW

File: service/controller/clusterapi/v29/cluster_resource_set.go [hunk 1]
```diff
@@ -531,14 +531,14 @@ func NewClusterResourceSet(config ClusterResourceSetConfig) (*controller.Resourc
...
 		machineDeploymentResource,
-		clusterAZsResource,
 		accountIDResource,
 		natGatewayAddressesResource,
 		peerRoleARNResource,
 		routeTableResource,
 		vpcCIDRResource,
 		tccpOutputsResource,
 		tccpSubnetResource,
+		clusterAZsResource,
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/cluster_resource_set.go]: Availability Zone inventory must be done only after TCCP subnets have been collected.

File: service/controller/clusterapi/v29/resource/tccp/template_render_test.go [hunk 2]
```diff
@@ -160,7 +161,23 @@ func defaultControllerContext() controllercontext.Context {
 				},
 			},
 			TenantCluster: controllercontext.ContextStatusTenantCluster{
-				AvailabilityZones:     []string{"eu-central-1a", "eu-central-1b", "eu-central-1c"},
+				AvailabilityZones: []controllercontext.ContextStatusTenantClusterAvailabilityZone{
+					{
+						Name:          "eu-central-1a",
+						PrivateSubnet: mustParseCIDR("10.100.3.0/27"),
+						PublicSubnet:  mustParseCIDR("10.100.3.32/27"),
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/resource/tccp/template_render_test.go]: These don't take effect on rendering outcome yet. That will be done in separate PR.

File: service/controller/clusterapi/v29/resource/tccpsubnet/resource.go [hunk 3]
```diff
@@ -81,6 +82,12 @@ func (r *Resource) addSubnetsToContext(ctx context.Context) error {
 						aws.String(cc.Status.TenantCluster.TCCP.VPC.ID),
 					},
 				},
+				{
+					Name: aws.String(key.TagTCCP),
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/resource/tccpsubnet/resource.go]: Filter out only control plane subnets from VPC (node pools also have subnets but they are part of separate CloudFormation stack & controller)

File: service/controller/clusterapi/v29/cluster_resource_set.go (hunk 1 above)
Comment (MarcelMue) [reply service/controller/clusterapi/v29/cluster_resource_set.go]: Have you checked if this ordering causes any unexpected problems? Is there a way to verify?

File: service/controller/clusterapi/v29/resource/clusterazs/create.go [hunk 4]
```diff
@@ -15,6 +20,18 @@ import (
...
+// subnetPair is temporary type for mapping existing subnets from
+// controllercontext to AZs.
+type subnetPair struct {
+	public  net.IPNet
+	private net.IPNet
+}
+
+func (sp subnetPair) areEmpty() bool {
+	return (sp.public.IP == nil && sp.public.Mask == nil) && (sp.private.IP == nil && sp.private.Mask == nil)
+
```
Comment (MarcelMue) [inline service/controller/clusterapi/v29/resource/clusterazs/create.go]:
```suggestion

```

File: service/controller/clusterapi/v29/resource/clusterazs/create.go [hunk 5]
```diff
@@ -48,30 +65,155 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
 }
+
+// fromEC2SubnetsToMap extracts availability zones and public / private subnet
+// CIDRs from given EC2 subnet slice and returns respectively structured map or
+// error on invalid data.
+func fromEC2SubnetsToMap(ss []*ec2.Subnet) (map[string]subnetPair, error) {
+	azMap := make(map[string]subnetPair)
+
+	for _, s := range ss {
+		if s == nil || s.AvailabilityZone == nil || s.CidrBlock == nil || s.Tags == nil {
```
Comment (MarcelMue) [inline service/controller/clusterapi/v29/resource/clusterazs/create.go]: Why do we continue here instead of error? Is there a case where this would be fine?

File: service/controller/clusterapi/v29/key/cluster.go [hunk 6]
```diff
@@ -36,11 +36,19 @@ const (
 	KubernetesSecurePort = 443
 )
 
+// MaximumNumberOfAZsInCluster defines the current limitation with Node Pools
+// implementation. Biggest limitation behind this is current IPAM
+// implementation that restricts network sizes. Another related problem is
+// restrictions in AWS resource structure.
+const MaximumNumberOfAZsInCluster = 4
```
Comment (MarcelMue) [inline service/controller/clusterapi/v29/key/cluster.go]: We should not have this as a magic number somewhere in the operator IMO. This will eventually cause confusion because someone can't find the reason why he/she can't add more node pools.

Comment (MarcelMue) [reply service/controller/clusterapi/v29/key/cluster.go]: Or when do we want to change this?

File: service/controller/clusterapi/v29/resource/tccpsubnet/resource.go (hunk 3 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/resource/tccpsubnet/resource.go]: Same worry here. Are we sure this can simply be changed? What is `cc.Status.TenantCluster.TCCP.Subnets` used for?

File: service/controller/clusterapi/v29/resource/clusterazs/create.go [hunk 7]
```diff
@@ -15,6 +20,18 @@ import (
...
+	public  net.IPNet
+	private net.IPNet
+}
+
+func (sp subnetPair) areEmpty() bool {
+	return (sp.public.IP == nil && sp.public.Mask == nil) && (sp.private.IP == nil && sp.private.Mask == nil)
+
+}
+
 func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/clusterazs/create.go]: The functionality here looks so complicated we should have unit tests for this to also see some input/output. I have really issues comprehending the code and what it does.

File: service/controller/clusterapi/v29/cluster_resource_set.go (hunk 1 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/cluster_resource_set.go]: I haven't given this PR a test run yet, but given I added this resource last week and it should be only used in `tccpResource` below, I'm optimistic here. Reason why I moved it here is that previous resource `tccpSubnetResource` is the one which fetches that `[]*ec2.Subnet` slice that this uses. Nevertheless, running this operator & creating a cluster would reveal problems.

File: service/controller/clusterapi/v29/key/cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tuommaki) [reply service/controller/clusterapi/v29/key/cluster.go]: I originally had this in resource package, but this is a constant that's needed for validation at least in `adapter` package and most probably elsewhere too so I thought it's less likely to define this multiple times when this is in central place, but I'm also fine moving it back to resource for now.

Comment (tuommaki) [reply service/controller/clusterapi/v29/key/cluster.go]: We must have this in this operator as it is an important piece of whole design of node pools infrastructure and behavior of dynamic AZs. I expect this to change when we can solve `Advanced IPAM` roadmap story and somewhere in the future possibly also refactor the TCCP structure further.

> because someone can't find the reason why he/she can't add more node pools.

Asynchronous CR reconciliation is not a design where one can provide reasonable feedback to end user. Somewhere in the future we might want to submit an event for these kind of cases, but that requires better error structuring. For end user UX there must be measures in place in `api`, `gsctl` and `happa`.

File: service/controller/clusterapi/v29/resource/clusterazs/create.go (hunk 7 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: Makes sense.

File: service/controller/clusterapi/v29/resource/clusterazs/create.go [hunk 8]
```diff
@@ -48,30 +65,155 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d MachineDeployments for tenant cluster", len(machineDeployments)))
 	}
 
-	var azs []string
+	var azs map[string]subnetPair
 	{
-		var azsMap map[string]struct{}
+		// Acquire AZs together with corresponding subnets from AWS EC2 API
+		// results.
+		azs, err = fromEC2SubnetsToMap(cc.Status.TenantCluster.TCCP.Subnets)
```
[the remark this answers is not in the document]
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: > Are we sure cc.Status.TenantCluster.TCCP.Subnets already holds the right information? 

We are sure that before CF stack creation it doesn't hold anything. At least that's the expectation.

> The code below to ensure the master subnet in there is a bit odd already. It would be nice to have once place to hold the subnets.

This code ensures that AZs configured for specific tenant cluster are assigned with correct subnets. When the AZ and corresponding subnet already exists in AWS, the connection must be kept. For rest of the AZs (the new ones that don't have resources in AWS yet) new subnet must be assigned from available ones.

> When this resource here is this place, I would question why we even need to use anything from cc.Status.TenantCluster.TCCP.Subnets.

Slice of AWS state of TCCP subnets is crucial for functionality of this code. This code wouldn't exist without and dynamic AZs wouldn't be possible without. Therefore it's needed.

File: service/controller/clusterapi/v29/resource/clusterazs/create.go [hunk 9]
```diff
@@ -48,30 +65,155 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
+				}
 			}
 		}
+	}
 
-		for az := range azsMap {
-			azs = append(azs, az)
+	var allocatedNetworks []net.IPNet
+	{
+		// Collect non-empty subnets from AZ-subnet -pairs.
```
[the remark this answers is not in the document]
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: Sure

File: service/controller/clusterapi/v29/resource/clusterazs/create.go [hunk 10]
```diff
@@ -48,30 +65,155 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
 	}
 
 	{
+		// Parse TCCP network CIDR.
+		_, clusterCIDR, err := net.ParseCIDR(key.StatusClusterNetworkCIDR(cr))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Split TCCP network between maximum number of AZs.
```
[the remark this answers is not in the document]
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: Yup

File: service/controller/clusterapi/v29/resource/clusterazs/create.go (hunk 5 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: My expectations towards that pile of pointers and input coming from outside of this resource are low. I don't see problem with skipping invalid entries here when functionality depends on actually gathered information.

Approach to these is always a double edged sword: If one errors here, it could be that for reason or another the slice contains valid entries that are just invalid in this use case and that prevents further reconciliation while everything would be fine. Alternatively we might skip them and there's possibility that everything goes fine but something gets omitted because of skipping something fishy here.

That being said, I'm fine returning error here if that makes sense for you.

File: service/controller/clusterapi/v29/resource/tccpsubnet/resource.go (hunk 3 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/tccpsubnet/resource.go]: Actually not. I also realized this filtering need to be moved to that mapping function above. These are also used for IPAM. I'll change this.

File: service/controller/clusterapi/v29/resource/clusterazs/create.go (hunk 5 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: My thinking would be that the data we manage internally should be considered complete. If information is missing the resource should be canceled early. Otherwise we should be able to expect perfect information at this point. Everything else is a bug that we need to fix and verify with unit tests.

File: service/controller/clusterapi/v29/key/cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/key/cluster.go]: I see. A note would do it for me that this should be moved elsewhere when the adapter package is gone.

File: service/controller/clusterapi/v29/resource/clusterazs/create.go (hunk 5 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: This returns an error now.

File: service/controller/clusterapi/v29/resource/clusterazs/create.go (hunk 7 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/clusterazs/create.go]: I split up the logic into smaller functions and wrote bunch of tests for them.

**Comment by tuommaki** [comment]
PTALA now that there are bunch of tests for this.

**Comment by tuommaki** [comment]
Successfully tested on `gauss`. Going with this now.
